content/training/MLaaS4HEP.md

```diff
@@ -0,0 +1,100 @@
+
+# both of aforementioned commands produce the following output
+Reading root://cms-xrd-global.cern.ch//store/data/Run2017F/Tau/NANOAOD/31Mar2018-v1/20000/6C6F7EAE-7880-E811-82C1-008CFA165F28.root
+# 1000 entries, 883 branches, 4.113945007324219 MB, 0.6002757549285889 sec, 6.853425235896175 MB/sec, 1.6659010326328503 kHz
+# 1000 entries, 883 branches, 4.067909240722656 MB, 1.3497390747070312 sec, 3.0138486148558896 MB/sec, 0.740883937302516 kHz
+###total time elapsed for reading + specs computing: 2.2570559978485107 sec; number of chunks 2
+###total time elapsed for reading: 1.9500117301940918 sec; number of chunks 2
+
+--- first pass: 1131872 events, (648-flat, 232-jagged) branches, 2463 attrs
+VMEM used: 29.896704 (MB) SWAP used: 0.0 (MB)
+<__main__.RootDataReader object at 0x7fb0cdfe4a00> init is complete in 2.265552043914795 sec
+Number of events  : 1131872
+# flat branches   : 648
+CaloMET_phi values in [-3.140625, 3.13671875] range, dim=N/A
+CaloMET_pt values in [0.783203125, 257.75] range, dim=N/A
+CaloMET_sumEt values in [820.0, 3790.0] range, dim=N/A
+```
+
+More examples about using uproot may be found [here](https://github.com/jpivarski/jupyter-talks/blob/master/2017-10-13-lpc-testdrive/uproot-introduction-evaluated.ipynb) and [here](https://github.com/jpivarski/jupyter-talks/blob/master/2017-10-13-lpc-testdrive/nested-structures-evaluated.ipynb).
+
+### How to train ML models on HEP ROOT data
+The training phase is managed by the `workflow.py` module which performs the following actions:
+- read all input ROOT files in chunks to compute a specs file
+- perform the training cycle (each time using a new chunk of events)
+  - create a new chunk of events taken proportionally from the input ROOT files
+    - extract and convert each event in a list of NumPy arrays
+    - normalize the events
+    - fix the Jagged Arrays dimension
+    - create the masking vector
+  - use the chunk to train the ML model provided by the user
+
+A schematic representation of the steps performed in the MLaaS4HEP pipeline, in particular those inside the Data Streaming and Data Training layers, is:
+![MLaaS4HEP-workflow](https://github.com/vkuznet/MLaaS4HEP/blob/master/images/mlaas4hep_workflow.png)
+
+This training procedure can be applied to a variety of use-cases but it should not be viewed as the only way to train data sets using the MLaaS4HEP framework. We left to the end user the final choice of ML strategy for concrete use-cases, where appropriate steps should be taken to check the convergence of the model, a proper set of metrics to monitor the training cycle, etc. If some ML algorithms don’t benefit much from splitting the data set in chunks, the user has the way to  customize the chunk size. In extreme cases the user can always fix the chunk size equal to the total number of events, restoring in this way the possibility to use the entire data set in one shot for the training process.

```

Comment:
1. From a formatting perspective, I think this text needs to stand out more. You can do something like:
```
> :warning: **<text_to_highlight>**
```
resulting in:
> :warning: **<text_to_highlight>**

Or any other way in markdown you know to make this stand out.

2. The other issue I have is that this paragraph, and the next, will not be so clear to someone who isn't already well versed in this issue of passing chunks or the entire dataset. Remember, most of CMS won't know that some models will converge with a subset of the data (a chunk) run through several epochs and some models will require all of the data run through the epochs. They also won't know about the *potential* for biasing the results to the last chunk of the data. So the statement should have: (1) an explanation of the various modes of running (i.e. chunking vs entire dataset), (2) an explanation of the drawbacks with each mode from a computing perspective (i.e. speed and ability to hold in memory), and (3) an explanation of the potential for bias towards the last chunk. Right now all that is said is to not use chunks if the model doesn't benefit. I also wouldn't consider it an extreme case to use the whole dataset per epoch rather than processing chunk-by-chunk as this is the normal procedure for Keras etc.

We appreciate the statement you've produced, but we would ask that you come up with a new version that addresses these concerns. We simply want to make sure that the general audience in CMS will be able to read this statement and understand their options.